### PR TITLE
Added minFee and silently skip already claimed prizes

### PR DESCRIPTION
**Pull request by asselstine**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -15,3 +15,4 @@
 [submodule "lib/pt-v5-vault"]
 	path = lib/pt-v5-vault
 	url = https://github.com/GenerationSoftware/pt-v5-vault
+	branch = gen-221-c4-issue-115
```
Comment (PierrickGT) [inline .gitmodules]: Remove

File: src/Claimer.sol [hunk 2]
```diff
@@ -51,35 +68,81 @@ contract Claimer is Multicall {
...
+      _vault,
+      _tier,
+      _winners,
+      _prizeIndices,
+      _feeRecipient,
+      feePerClaim
     );
+  }
+
+  function _computeFeePerClaimForBatch(
```
Comment (PierrickGT) [inline src/Claimer.sol:124]: Add Natspec doc.

File: src/Claimer.sol [hunk 3]
```diff
@@ -51,35 +68,81 @@ contract Claimer is Multicall {
...
-    vault.claimPrizes(tier, winners, prizeIndices, feePerClaim, _feeRecipient);
+    return _computeFeePerClaim(
+      _computeMaxFee(_tier, prizePool.numberOfTiers()),
+      claimCount,
+      prizePool.claimCount()
+    );
+  }
 
-    return feePerClaim * claimCount;
+  function _claim(
```
Comment (PierrickGT) [inline src/Claimer.sol:150]: Add Natspec doc.

File: src/Claimer.sol [hunk 4]
```diff
@@ -51,35 +68,81 @@ contract Claimer is Multicall {
...
-    uint8 tier,
-    address[] calldata winners,
-    uint32[][] calldata prizeIndices,
-    address _feeRecipient
+    Vault _vault,
+    uint8 _tier,
+    address[] calldata _winners,
+    uint32[][] calldata _prizeIndices,
+    address _feeRecipient,
+    uint256 _minVrgdaFeePerClaim
```
Comment (PierrickGT) [inline src/Claimer.sol:97]: I don't understand this parameter.
If I provide `_minVrgdaFeePerClaim` when calling `claimPrizes`, can't I just pass `0` and the condition `(feePerClaim < _minVrgdaFeePerClaim)` will always be false?

File: src/Claimer.sol [hunk 5]
```diff
@@ -51,35 +68,81 @@ contract Claimer is Multicall {
...
     );
+  }
+
+  function _computeFeePerClaimForBatch(
+    uint8 _tier,
+    address[] calldata _winners,
+    uint32[][] calldata _prizeIndices
+  ) internal view returns (uint256) {
+    uint256 claimCount;
+    for (uint i = 0; i < _winners.length; i++) {
```
Comment (PierrickGT) [inline src/Claimer.sol]: Cache `_winners.length` in a variable.

File: src/Claimer.sol [hunk 6]
```diff
@@ -51,35 +68,81 @@ contract Claimer is Multicall {
...
+    uint32[][] calldata _prizeIndices,
+    address _feeRecipient,
+    uint96 _feePerClaim
+  ) internal returns (uint256) {
+    uint256 actualClaimCount;
+    uint256 winnersLength = _winners.length;
+    for (uint256 w = 0; w < winnersLength; w++) {
+      uint256 prizeIndicesLength = _prizeIndices[w].length;
+      for (uint256 p = 0; p < prizeIndicesLength; p++) {
+        if (0 < _vault.claimPrize(
```
Comment (PierrickGT) [inline src/Claimer.sol]: This condition is a bit confusing. I would have used `!=` instead.
```suggestion
        if (0 != _vault.claimPrize(
```

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: src/Claimer.sol (hunk 4 above)
Comment (asselstine) [reply src/Claimer.sol:97]: Yes that's right; if the minimum is zero then it'll always work just fine.

Does that make sense?

**Pull request merged**